pkg/controller/egress/validate_test.go

```diff
@@ -0,0 +1,206 @@
+// limitations under the License.
+
+package egress
+
+import (
+	"encoding/json"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	admv1 "k8s.io/api/admission/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+
+	crdv1alpha2 "antrea.io/antrea/pkg/apis/crd/v1alpha2"
+)
+
+func marshal(object runtime.Object) []byte {
+	raw, _ := json.Marshal(object)
+	return raw
+}
+
+func TestEgressControllerValidateExternalIPPool(t *testing.T) {
+	tests := []struct {
+		name             string
+		request          *admv1.AdmissionRequest
+		expectedResponse *admv1.AdmissionResponse
+	}{
+		{
+			name: "CREATE operation should be allowed",
+			request: &admv1.AdmissionRequest{
+				Name:      "foo",
+				Operation: "CREATE",
+				Object:    runtime.RawExtension{Raw: marshal(newExternalIPPool("foo", "10.10.10.0/24", "", ""))},
+			},
+			expectedResponse: &admv1.AdmissionResponse{Allowed: true},
+		},
+		{
+			name: "Deleting IPRange should not be allowed",
+			request: &admv1.AdmissionRequest{
+				Name:      "foo",
+				Operation: "UPDATE",
+				OldObject: runtime.RawExtension{Raw: marshal(newExternalIPPool("foo", "10.10.10.0/24", "10.10.20.1", "10.10.20.2"))},
+				Object:    runtime.RawExtension{Raw: marshal(newExternalIPPool("foo", "10.10.10.0/24", "", ""))},
+			},
+			expectedResponse: &admv1.AdmissionResponse{
+				Allowed: false,
+				Result: &metav1.Status{
+					Message: "existing IPRanges [10.10.20.1-10.10.20.2] cannot be deleted",
+				},
+			},
+		},
+		{
+			name: "Adding IPRange should be allowed",
+			request: &admv1.AdmissionRequest{
+				Name:      "foo",
+				Operation: "UPDATE",
+				OldObject: runtime.RawExtension{Raw: marshal(newExternalIPPool("foo", "10.10.10.0/24", "", ""))},
+				Object:    runtime.RawExtension{Raw: marshal(newExternalIPPool("foo", "10.10.10.0/24", "10.10.20.1", "10.10.20.2"))},
+			},
+			expectedResponse: &admv1.AdmissionResponse{Allowed: true},
+		},
+		{
+			name: "DELETE operation should be allowed",
+			request: &admv1.AdmissionRequest{
+				Name:      "foo",
+				Operation: "DELETE",
+				Object:    runtime.RawExtension{Raw: marshal(newExternalIPPool("foo", "10.10.10.0/24", "", ""))},
+			},
+			expectedResponse: &admv1.AdmissionResponse{Allowed: true},
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			c := newController(nil, nil)
+			review := &admv1.AdmissionReview{
+				Request: tt.request,
+			}
+			gotResponse := c.ValidateExternalIPPool(review)
+			assert.Equal(t, tt.expectedResponse, gotResponse)
+		})
+	}
+}
+
+func TestEgressControllerValidateEgress(t *testing.T) {
+	tests := []struct {
+		name                   string
+		existingExternalIPPool *crdv1alpha2.ExternalIPPool
+		request                *admv1.AdmissionRequest
+		expectedResponse       *admv1.AdmissionResponse
+	}{
+		{
+			name:                   "CREATE: Requesting IP from non-existing ExternalIPPool should not be allowed",
```

Comment:
why the weird formatting for these testcase names?

Reply:
I wanted to make names different for CREATE and UPDATE operations when both of them are requesting invalid IPs.
I have updated the UPDATE case to "Updating to ...".